File: frontend/src/services/authService.ts
import { api } from './api'
import type {
  LoginRequest,
  RegisterRequest,
  TokenResponse,
  RefreshTokenRequest,
  User,
} from '../types'

/**
 * Authentication service for managing user authentication operations
 */
class AuthService {
  private readonly AUTH_BASE_URL = '/auth'

  /**
   * Register a new user
   */
  async register(data: RegisterRequest): Promise<TokenResponse> {
    const response = await api.post<TokenResponse>(`${this.AUTH_BASE_URL}/register`, data)
    return response.data
  }

  /**
   * Login with email and password
   */
  async login(data: LoginRequest): Promise<TokenResponse> {
    const response = await api.post<TokenResponse>(`${this.AUTH_BASE_URL}/login`, data)
    return response.data
  }

  /**
   * Logout from current session
   */
  async logout(): Promise<void> {
    const refreshToken = localStorage.getItem('refresh_token')
    if (refreshToken) {
      try {
        await api.post(`${this.AUTH_BASE_URL}/logout`, { refresh_token: refreshToken })
      } catch (error) {
        // Ignore errors during logout
        console.error('Logout error:', error)
      }
    }
    this.clearTokens()
  }

  /**
   * Logout from all sessions
   */
  async logoutAll(): Promise<void> {
    try {
      await api.post(`${this.AUTH_BASE_URL}/logout-all`)
    } catch (error) {
      console.error('Logout all error:', error)
    }
    this.clearTokens()
  }

  /**
   * Refresh access token using refresh token
   */
  async refreshToken(data: RefreshTokenRequest): Promise<TokenResponse> {
    const response = await api.post<TokenResponse>(`${this.AUTH_BASE_URL}/refresh`, data)
    return response.data
  }

  /**
   * Get current authenticated user
   */
  async getCurrentUser(): Promise<User> {
    const response = await api.get<User>(`${this.AUTH_BASE_URL}/me`)
    return response.data
  }

  /**
   * Store tokens in localStorage
   */
  storeTokens(accessToken: string, refreshToken: string): void {
    localStorage.setItem('access_token', accessToken)
    localStorage.setItem('refresh_token', refreshToken)
  }

  /**
   * Clear all stored tokens
   */
  clearTokens(): void {
    localStorage.removeItem('access_token')
    localStorage.removeItem('refresh_token')
  }

  /**
   * Get stored access token
   */
  getAccessToken(): string | null {
    return localStorage.getItem('access_token')
  }

  /**
   * Get stored refresh token
   */
  getRefreshToken(): string | null {
    return localStorage.getItem('refresh_token')
  }

  /**
   * Check if user is authenticated (has access token)
   */
  isAuthenticated(): boolean {
    return !!this.getAccessToken()
  }
}

export const authService = new AuthService()

File: backend/app/services/auth_service.py
"""Authentication service for user registration, login, and token management"""

from datetime import datetime, timedelta, timezone
from typing import Optional, Tuple

from jose import JWTError
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.exceptions import UnauthorizedException
from app.core.security import (
    create_access_token,
    create_refresh_token,
    decode_token,
    get_password_hash,
    verify_password,
    verify_token_type,
)
from app.db.models import User, UserSession
from app.repositories import UserRepository, UserSessionRepository
from app.schemas import TokenResponse, UserCreate, UserLogin, UserResponse


class AuthService:
    """Service for authentication operations"""

    def __init__(self, session: AsyncSession):
        """Initialize service with database session"""
        self.session = session
        self.user_repo = UserRepository(session)
        self.session_repo = UserSessionRepository(session)

    async def register_user(
        self,
        user_data: UserCreate,
        ip_address: Optional[str] = None,
        user_agent: Optional[str] = None,
    ) -> TokenResponse:
        """
        Register new user and return authentication tokens

        Args:
            user_data: User registration data
            ip_address: Client IP address
            user_agent: Client user agent

        Returns:
            TokenResponse with access/refresh tokens and user data

        Raises:
            ValueError: If email already exists
        """
        # Check if user already exists
        existing_user = await self.user_repo.get_by_email(user_data.email)
        if existing_user:
            raise ValueError(f"Email {user_data.email} is already registered")

        # Create new user
        hashed_password = get_password_hash(user_data.password)
        new_user = User(
            email=user_data.email,
            password_hash=hashed_password,
            name=user_data.name,
            subscription_tier="free",
        )

        await self.user_repo.create(new_user)
        await self.session.commit()

        # Create tokens and session
        return await self._create_user_session(
            new_user, ip_address=ip_address, user_agent=user_agent
        )

    async def authenticate_user(
        self,
        credentials: UserLogin,
        ip_address: Optional[str] = None,
        user_agent: Optional[str] = None,
    ) -> TokenResponse:
        """
        Authenticate user with email and password

        Args:
            credentials: User login credentials
            ip_address: Client IP address
            user_agent: Client user agent

        Returns:
            TokenResponse with access/refresh tokens and user data

        Raises:
            UnauthorizedException: If credentials are invalid
        """
        # Get user by email
        user = await self.user_repo.get_by_email(credentials.email)
        if not user:
            raise UnauthorizedException("Invalid email or password")

        # Verify password
        if not verify_password(credentials.password, user.password_hash):
            raise UnauthorizedException("Invalid email or password")

        # Update last login
        user.update_last_login()
        await self.user_repo.update(user)
        await self.session.commit()

        # Create tokens and session
        return await self._create_user_session(
            user, ip_address=ip_address, user_agent=user_agent
        )

    async def refresh_access_token(
        self, refresh_token: str
    ) -> Tuple[str, UserResponse]:
        """
        Refresh access token using valid refresh token

        Args:
            refresh_token: JWT refresh token

        Returns:
            Tuple of (new access token, user data)

        Raises:
            UnauthorizedException: If refresh token is invalid or expired
        """
        # Verify token type
        if not verify_token_type(refresh_token, "refresh"):
            raise UnauthorizedException("Invalid refresh token type")

        # Get session from database
        user_session = await self.session_repo.get_by_refresh_token(refresh_token)
        if not user_session:
            raise UnauthorizedException("Invalid refresh token")

        # Check if session is valid
        if not user_session.is_valid:
            raise UnauthorizedException("Refresh token has been revoked or expired")

        # Get user
        user = await self.user_repo.get_by_id(user_session.user_id)
        if not user:
            raise UnauthorizedException("User not found")

        # Update session last accessed
        user_session.update_last_accessed()
        await self.session_repo.create(user_session)
        await self.session.commit()

        # Create new access token
        access_token = create_access_token(user.id)
        user_response = UserResponse.model_validate(user)

        return access_token, user_response

    async def logout(self, refresh_token: str) -> bool:
        """
        Logout user by revoking refresh token

        Args:
            refresh_token: JWT refresh token to revoke

        Returns:
            True if token was revoked, False if not found
        """
        success = await self.session_repo.revoke_by_refresh_token(refresh_token)
        if success:
            await self.session.commit()
        return success

    async def logout_all_sessions(self, user_id: int) -> int:
        """
        Logout user from all devices by revoking all sessions

        Args:
            user_id: User ID

        Returns:
            Number of sessions revoked
        """
        count = await self.session_repo.revoke_all_user_sessions(user_id)
        await self.session.commit()
        return count

    async def get_user_by_id(self, user_id: int) -> Optional[User]:
        """Get user by ID"""
        return await self.user_repo.get_by_id(user_id)

    async def verify_access_token(self, token: str) -> User:
        """
        Verify access token and return user

        Args:
            token: JWT access token

        Returns:
            User object

        Raises:
            UnauthorizedException: If token is invalid or user not found
        """
        try:
            # Decode and verify token
            payload = decode_token(token)

            # Verify token type
            if payload.get("type") != "access":
                raise UnauthorizedException("Invalid token type")

            # Extract user ID
            user_id_str = payload.get("sub")
            if not user_id_str:
                raise UnauthorizedException("Invalid token payload")

            user_id = int(user_id_str)

            # Get user from database
            user = await self.user_repo.get_by_id(user_id)
            if not user:
                raise UnauthorizedException("User not found")

            return user

        except JWTError as e:
            raise UnauthorizedException(f"Invalid token: {str(e)}") from e
        except ValueError as e:
            raise UnauthorizedException(f"Invalid user ID: {str(e)}") from e

    async def _create_user_session(
        self,
        user: User,
        ip_address: Optional[str] = None,
        user_agent: Optional[str] = None,
    ) -> TokenResponse:
        """Create new session with tokens for user"""
        # Create tokens
        access_token = create_access_token(user.id)
        refresh_token = create_refresh_token(user.id)

        # Create session in database
        expires_at = datetime.now(timezone.utc) + timedelta(days=30)
        user_session = UserSession(
            user_id=user.id,
            refresh_token=refresh_token,
            ip_address=ip_address,
            user_agent=user_agent,
            expires_at=expires_at,
        )

        await self.session_repo.create(user_session)
        await self.session.commit()

        # Create response
        user_response = UserResponse.model_validate(user)
        return TokenResponse(
            access_token=access_token,
            refresh_token=refresh_token,
            user=user_response,
        )
